Merge duality neutron

jcompagni10 opened the pull request:
This PR has been moved here https://github.com/neutron-org/neutron/pull/341 so as to use a branch from the neutron org
but keeping this one around to preserve comment history. 

Merge duality modules (dex, epochs, gmp, ibcswap & incentives) into neutron.

Commit:
copy over duality files

Commit:
update duality code to connect with neutron

Commit:
downgrade packet-forward-middleware

the version of packet-forward-middleware being used by neutron breaks duality's
swap-and-forward middleware due to a change in how receiever addresses are
handled.

swelf19 commented:
I can see its a draft yet, but a couple of sugestions.
1) Pls add new stores (dextypes.StoreKey, incentivestypes.StoreKey, epochstypes.StoreKey) to the upgrade handler https://github.com/jcompagni10/neutron/blob/merge_duality_neutron/app/upgrades/nextupgrade/constants.go#L23
2) we are moving every module params from a paramsKeeper to the module's own store, since params keeper is deprecated. Pls follow the convention

Commit:
add dex, incentives and epochs modules to storeUpgrades keys

jcompagni10 commented:
> 1. Pls add new stores (dextypes.StoreKey, incentivestypes.StoreKey, epochstypes.StoreKey) to the upgrade handler https://github.com/jcompagni10/neutron/blob/merge_duality_neutron/app/upgrades/nextupgrade/constants.go#L23

@swelf19 just want to confirm this this satisfies this above https://github.com/neutron-org/neutron/pull/312/commits/346d8e543200540a209edfb4cb10747f1bb85c1b

Commit:
change dex params to kv store

Commit:
edit and recompile protos to match neutron namespace

Commit:
switch incentives to kvstore params

jcompagni10 commented:
> 2. we are moving every module params from a paramsKeeper to the module's own store, since params keeper is deprecated. Pls follow the convention

See https://github.com/jcompagni10/neutron/commit/757478d30f58249ac19a754a85e5373d2708fcd2 & https://github.com/jcompagni10/neutron/commit/ffe7bdba58db22bbaca0adeea77f58429ee50ede

Commit:
start using "cosmossdk.io/math"

Commit:
FIX: deprecation github.com/cosmos/cosmos-sdk/types/errors => cosmossdk.io/errors

Commit:
FIX: missing import

Commit:
misc linting fixes

Commit:
go fumpt --extra on all duality files

swelf19 commented:
1) starting brand new node with `make init` fails with `panic: BlockGas meter must be initialized. Genesis must provide value for Block.MaxGas` pls fix it either by configuring neutron to init proper genesis or by modifying genesis in `init-neutrond.sh` script
2) pls regenerate docs/static/openapi.yml (`ignite generate openapi`)
3) fix conflicts

Commit:
General Test cleanup

cleanup KeeperTestHelper and re-use for dex testing

centralize event testing assertions

rename test suite to be more informative

Commit:
set max_gas in init-neutrond.sh

duality code requires max gas to be set initChain panics

Commit:
remove vestigial code

Commit:
rename duality => neutron where applicable in code

jcompagni10 commented:
> 1. starting brand new node with `make init` fails with `panic: BlockGas meter must be initialized. Genesis must provide value for Block.MaxGas` pls fix it either by configuring neutron to init proper genesis or by modifying genesis in `init-neutrond.sh` script

Resolved
> 2. pls regenerate docs/static/openapi.yml (`ignite generate openapi`)

What version of ignite are you using to generate openapi. I just ran it with ignite v0.27.1 and it made significant changes to the current openapi.yml unrelated to the duality merge
> 3. fix conflicts

resolved

Also resolved all linting. And did a more general cleanup of the duality code.

NeverHappened commented on app/app.go:
```diff
@@ -809,6 +895,10 @@ func New(
 		crontypes.ModuleName,
 		globalfee.ModuleName,
+		ibcswaptypes.ModuleName,
+		dextypes.ModuleName,
```
why no have the same order as in endblocker? or is this intentional?

NeverHappened commented on app/upgrades/nextupgrade/constants.go:
```diff
@@ -21,6 +24,9 @@ var Upgrade = upgrades.Upgrade{
 			consensusparamtypes.ModuleName,
 			crisistypes.ModuleName,
 			auctiontypes.ModuleName,
+			dextypes.ModuleName,
```
why is there no ibcswap module added?

jcompagni10 replied:
ibcswap does not maintain a KVstore, so if I understand correctly it should not have to be used in any migration. Correct me if I'm wrong. @NeverHappened

NeverHappened commented on x/dex/client/cli/query.go:
```diff
@@ -0,0 +1,50 @@
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	// "github.com/cosmos/cosmos-sdk/client/flags"
+	// sdk "github.com/cosmos/cosmos-sdk/types"
```
why commented?

NeverHappened commented on x/dex/client/cli/query.go:
```diff
@@ -0,0 +1,50 @@
+
+import (
+	"fmt"
+	// "strings"
```
why commented?

sotnikov-s commented on app/app.go:
```diff
@@ -234,6 +256,8 @@ var (
 		ccvconsumertypes.ConsumerToSendToProviderName: nil,
 		tokenfactorytypes.ModuleName:                  {authtypes.Minter, authtypes.Burner},
 		crontypes.ModuleName:                          nil,
+		dextypes.ModuleName:                           {authtypes.Minter, authtypes.Burner, authtypes.Staking},
```
there is no staking module in Neutron. probably some leftover `authtypes.Staking` permission requirement? also I can see `minter` and `burner` permissions usage, but no staking-related stuff

sotnikov-s commented on app/app.go:
```diff
@@ -604,6 +632,57 @@ func New(
+		appCodec,
+		keys[dextypes.StoreKey],
+		keys[dextypes.MemStoreKey],
+		app.BankKeeper,
```
I don't think it's a good idea to give ultimate mint permissions for a module. IMO we should have a [mintCoinsRestrictionFn](https://github.com/neutron-org/cosmos-sdk/blob/fcccb9a3162510de3b39c0253f22918cbe56360d/x/bank/keeper/keeper.go#L60) defined for the dex module e.g. like we have for the [tokenfactory module](https://github.com/jcompagni10/neutron/blob/87da06b389b810ac85a2353824bf1a5d0761aaf3/app/app.go#L614) which only allows the module to mint tokens of specific format (`factory/{creator}/{subdenom}`). in your case it could be `neutron/pool/{pool_id}`

sotnikov-s commented on x/ibcswap/types/expected_keepers.go:
```diff
@@ -0,0 +1,14 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// BankKeeper defines the expected interface that the swap middleware needs in order to facilitate refunds.
+type BankKeeper interface {
+	SendCoins(ctx sdk.Context, fromAddr, toAddr sdk.AccAddress, amt sdk.Coins) error
+	MintCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	SendCoinsFromModuleToAccount(ctx sdk.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins) error
+	SendCoinsFromAccountToModule(ctx sdk.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins) error
+}
```
some of the methods aren't used, please clean up

jcompagni10 replied:
👍

sotnikov-s commented on x/ibcswap/keeper/keeper.go:
```diff
@@ -0,0 +1,189 @@
+		}
+
+		// burn the coins
+		err = k.bankKeeper.BurnCoins(ctx, transfertypes.ModuleName, sdk.NewCoins(token))
```
why is this call made on behalf of the ibc transfer module? probably a misprint. if no particular reason to do so, please consider giving explicit burn permissions to the ibcswap module and making a clean conscious call here

sotnikov-s commented on x/dex/types/expected_keepers.go:
```diff
@@ -0,0 +1,17 @@
+	MintCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error
+	IterateAccountBalances(ctx sdk.Context, addr sdk.AccAddress, cb func(sdk.Coin) bool)
+	SpendableCoins(ctx sdk.Context, addr sdk.AccAddress) sdk.Coins
```
`SpendableCoins` is not used

sotnikov-s commented on app/app.go:
```diff
@@ -1243,6 +1354,9 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	paramsKeeper.Subspace(tokenfactorytypes.StoreKey).WithKeyTable(tokenfactorytypes.ParamKeyTable())
 	paramsKeeper.Subspace(interchainqueriesmoduletypes.StoreKey).WithKeyTable(interchainqueriesmoduletypes.ParamKeyTable())
 	paramsKeeper.Subspace(interchaintxstypes.StoreKey).WithKeyTable(interchaintxstypes.ParamKeyTable())
+	paramsKeeper.Subspace(dextypes.ModuleName)
+	paramsKeeper.Subspace(epochstypes.ModuleName)
+	paramsKeeper.Subspace(incentivestypes.ModuleName)
```
why no key table used for modules' params?

sotnikov-s commented on network/init-neutrond.sh:
```diff
@@ -643,6 +643,7 @@ set_genesis_param_jq ".app_state.globalfee.params.bypass_min_fee_msg_types" "$BY
 set_genesis_param proposer_fee                "\"0.25\""                                    # builder(POB)
 set_genesis_param escrow_account_address      "\"$DAO_CONTRACT_ADDRESS_B64\","              # builder(POB)
 set_genesis_param sudo_call_gas_limit                "\"1000000\""                          # contractmanager
+set_genesis_param max_gas                      "\"40000000\""                               # consensus_params
```
is this value approved or something? @swelf19 @pr0n00gler

jcompagni10 replied:
The dex module requires max gas to be set to avoid certain DOS scenarios wrt purging expired limit orders. We settled on this number as a sensible max gas per block (ie. high enough to not be hit under normal conditions, low enough to prevent DOS) but we can certainly change it if you guys have an opinion.

jcompagni10 replied:
@swelf19 @sotnikov-s @pr0n00gler

mmikev replied:
Currently we have `max_gas` equals to 30M. I prefer leave it as it is

mmikev replied:
Changed it to 1B since it's required for our integration tests. On mainnet 'max_gas' equals to 30M

sotnikov-s commented on wasmbinding/stargate_allowlist.go:
```diff
@@ -47,5 +50,30 @@ func AcceptedStargateQueries() wasmkeeper.AcceptedStargateQueries {
+		"/neutron.dex.Query/EstimateMultiHopSwap":      &dextypes.QueryEstimateMultiHopSwapResponse{},
+		"/neutron.dex.Query/EstimatePlaceLimitOrder":   &dextypes.QueryEstimatePlaceLimitOrderResponse{},
+
+		// incentives
+		"/neutron.incentives.Query/GetModuleStatus":         &incentivestypes.GetModuleStatusResponse{},
+		"/neutron.incentives.Query/GetGaugeByID":            &incentivestypes.GetGaugeByIDResponse{},
+		"/neutron.incentives.Query/GetGauges":               &incentivestypes.GetGaugesResponse{},
+		"/neutron.incentives.Query/GetStakeByID":            &incentivestypes.GetStakeByIDResponse{},
+		"/neutron.incentives.Query/GetStakes":               &incentivestypes.GetStakesResponse{},
+		"/neutron.incentives.Query/GetFutureRewardEstimate": &incentivestypes.GetFutureRewardEstimateResponse{},
+		"/neutron.incentives.Query/GetAccountHistory":       &incentivestypes.GetAccountHistoryResponse{},
+		"/neutron.incentives.Query/GetGaugeQualifyingValue": &incentivestypes.GetGaugeQualifyingValueResponse{},
```
you probably've forgotten to add module's params query here?

jcompagni10 replied:
GetModuleStatus includes the params, but I can change it if you prefer @sotnikov-s
